### DISCUSSION TITLE
Time to set inode_cache enabled by default?

### DISCUSSION THREAD

**Discussion by liljenzin** [ccache/ccache#1086] [Development]
The inode cache has been in the master branch for more than two years now, though disabled by default. So far I have not heard about any cases where the inode cache has caused incorrectness or slower performance than when running without it.

The inode cache is a cache for source file hashes with the purpose to avoid hashing the same header files again and again for each compilation unit they are included in, based on the assumption that content of a named file cannot have changed (and thus not the file hash) unless inode number, device number and/or time stamps have changed since the hash was last computed. As the cache resides in a memory mapped file it can be accessed simultaneously from concurrent ccache processes on the same machine without measurable performance overhead. The cache is currently available on Linux and other POSIX systems.

What to expect from it?

When building small projects where each source file only includes a few other files, using the inode cache will not make a big difference since the time ccache spends on computing hash values is small already without the inode cache. It will not be slower but also not much faster. When building larger projects, such as the chromium browser, a lot of source files include thousands of other source files (since header files usually include other header files recursively), which makes ccache spending most of its time in computing file hashes. The use of the inode cache eliminates most of the hashing. In the chromium case this reduces the cpu time spent in ccache by about 50%, which makes the total time for building the Chrome browser from a hot ccache about 10% faster (not twice as fast since the build script is doing more than compiling C/C++ though).

Would be good to hear what experiences other people have, if anyone is using the inode cache at all, and if there are reasons to not enable it by default in some future major ccache release?

**Comment by jrosdahl** [comment]
I have the inode cache enabled myself for dogfooding purposes and I haven't noticed any problems. On the other hand, I haven't heard about anybody else using it so I guess it's hard to draw any confident conclusions. That said, I think that it would be OK to enable the inode cache by default now.

Related: Some OSes, at least Gentoo, [patch ccache to not use `/run/user/$UID`](https://gitweb.gentoo.org/repo/gentoo.git/tree/dev-util/ccache/files/ccache-4.6-avoid-run-user.patch?id=035fe04cb6f6f041330ab7561bdd1dc7687343de) as the temporary directory. I've also had to make a handful of bug fixes related this: a0edd4294f6a5a2d3f0c7b01273736f975f250e1 ef2e922f9642f943199138447b29ec53fa63ea68 f2ff4b90b8767ea0bc2d49162d85d5f08d72f24c ce6b0c3dbdf1f84ab625d4c86770297f93b87032. Not sure if it's the last one -- it is apparently a bug attractor. Maybe we should also consider reverting to always use `$CCACHE_DIR/tmp` as the default temporary directory? If I remember correctly, one reason for using `/run/user/$UID` was to avoid failing tests on NFS, but I think it would be perfectly fine to simply not run the tests on NFS. What do you think?

**Reply by liljenzin** [reply]
If not using a default location that is local by specification (such as /run/user/) we need to check that it is local each time ccache is run since the same path could be local on one host (the nfs server) and remote on other hosts (the nfs clients), thus not enough to check when the configuration is changed. Otherwise we could end up in data corruption since mutexes will not work as supposed.

Perhaps it is good enough to do add a check in Linux, otherwise each UNIX flavor appears to have its own way of checking if a filesystem is remote or not.

**Reply by jrosdahl** [reply]
> If not using a default location that is local by specification (such as /run/user/) we need to check that it is local each time ccache is run since the same path could be local on one host (the nfs server) and remote on other hosts (the nfs clients)

Isn't that already done by https://github.com/ccache/ccache/blob/561be2085df94b3c35dd803e56668e6feafc93b7/src/InodeCache.cpp#L138-L143? On the other hand:

> Perhaps it is good enough to do add a check in Linux, otherwise each UNIX flavor appears to have its own way of checking if a filesystem is remote or not.

Right. However, just checking if the file system is NFS doesn't feel good enough since it could potentially be, say, some SMB or FUSE-based file system. Especially if somebody with the [the Gentoo patch](https://gitweb.gentoo.org/repo/gentoo.git/tree/dev-util/ccache/files/ccache-4.6-avoid-run-user.patch?id=035fe04cb6f6f041330ab7561bdd1dc7687343de) puts the cache on such a file system. Is there a better way? Maybe we can have an allowlist of acceptable file systems instead? What do you think?

**Reply by liljenzin** [reply]
Perhaps we should only make inode cache default enabled on Linux this time since we are not checking filesystem type on other systems yet? Otherwise we could break existing configurations that worked before the change. And perhaps only white list filesystems that are known to work and then add others if requested by users?

**Reply by jrosdahl** [reply]
Yes, that was my idea as well. I'll look into it.

**Reply by jrosdahl** [reply]
Done: 5edcc6c36d881e0778c8d6b0b430a1ca054426d8. Enabled inode cache by default: bf935e13de26cc6cea78b100fd3ebaf6dbef5fc5.

**Reply by thesamesam** [reply]
Heh, I keep wanting to drop that Gentoo patch (as one of the maintainers of ccache in Gentoo) but last time I tried, things blew up and I haven't had time to go back to it (or even figure out where the problem is, as I inherited the patch).

**Comment by jrosdahl** [comment]
@liljenzin: I've now tested to enable the inode cache by default and also rereviewed the implementation. I found one bug and two minor inconveniences:

1. Bug: InodeCache caches calls to `hash_source_code_file` and `hash_binary_file`, which works well for the common case but not when there is a `__DATE__` or `__TIMESTAMP__` in the code. This is because `hash_source_code_string` (called from `hash_source_code_file` via `hash_source_code_file_nocache`) on an inode cache miss [adds time information to the hash](https://github.com/ccache/ccache/blob/1ee5365507ff3dd3d071f9ccd9380ad5c661ee12/src/hashutil.cpp#L250-L307) in case one of those macros were found. On an inode cache hit, the return value of `hash_source_code_file` will be correctly fetched from the cache, but the hash sum will only be updated with (the hash of) the include file and not the time information. (This can be reproduced by enabling the inode cache and running the source_date_epoch test.)
2. Minor inconvenience: Hashes stored in the direct mode manifest are different from those produced by `ccache --hash-file` since `hash_source_code_file` [hashes the file digest](https://github.com/ccache/ccache/blob/1ee5365507ff3dd3d071f9ccd9380ad5c661ee12/src/hashutil.cpp#L347) which can make debugging a bit trickier.
3. Minor inconvenience: Hashes stored in the direct mode manifest will differ depending on whether the inode cache is enabled since `hash_source_code_file` hashes the file hash digest if the inode cache is enabled and hashes the file content directly if not. This inconsistency can make debugging a bit trickier.

Other than that I think it works very well. :slightly_smiling_face:

**Reply by jrosdahl** [reply]
I've tried to address these issues in 561be2085df94b3c35dd803e56668e6feafc93b7. @liljenzin: A review of the change would be appreciated.

However, when enabling the inode cache by default, the direct and sanitize_blacklist tests fail on macOS more often than not. The reason is that commands like https://github.com/ccache/ccache/blob/561be2085df94b3c35dd803e56668e6feafc93b7/test/suites/sanitize_blacklist.bash#L75 (often but not always) don't change the inode mtime/ctime (at least not quickly enough), so there will be a false positive inode cache hit on the next compilation! That's...surprising. Here's a test run with debug output that shows the error: https://github.com/jrosdahl/ccache/runs/7647301688

I haven't looked into this much except that I checked that [the file system used in the GitHub macOS environment is APFS](https://github.com/jrosdahl/ccache/runs/7647301688#step:4:542). If this happens on macOS APFS, can it happen on other systems as well?

**Reply by liljenzin** [reply]
The pushed changes look good to me.

Regarding time resolution on macOS and APFS breaking ccache.
Consider the following Makefile:
```
bar: bar.c
	gcc bar.c -o bar

bar.c: foo.c
	sed s/foo/bar/g foo.c > bar.c
```
If bar.c can be changed without the timestamp getting updated, then make will conclude everything is up to date and nothing needs to be recompiled. So a system where this can happen can not be a reliable build platform at all, regardless if ccache is used or not.

**Reply by jrosdahl** [reply]
I thought that the problem isn't due to file system time resolution since APFS has nanosecond resolution and we use the subsecond time fields in `struct stat`. But wait, now I get it... We read the `st_Xtim` fields, but it's [`st_Xtimespec` in `struct stat` on macOS](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/stat.2.html). Fixed: bcdaeb5c6082717bac984ab2ffafd0bdee552458

**Reply by liljenzin** [reply]
Good you found it.

**Comment by jdscriven** [comment]
Hi, I wanted share our recent experience with enabling inode cache in ccache 4.2 on a very large codebase, to a very large set of users on Linux (RHEL8). Our main motivation for using inode cache was our use of a large, statically linked clang (100MB+), combined with compiler_check=content, which meant we were spending a very large amount of time hashing this 100MB file for every compile. We do roughly 50,000 compiles during a typical build and inode cache reduced our average (ccache hit) compile from 118ms to 14ms.

Other than the big win in compile times, enabling inode cache went largely unnoticed, which is what we want. There were a few issues reported, that all traced back to our use of ccache when running autoconf configure scripts. configure generates many incantations of conftest.c, each with a different compiler test. These files are fairly quick to generate and compile, and we had some users still using ext3 with 1 second resolution timestamps. We found that conftest.c would get generated and compiled and then get deleted, generated and compiled again in the same second and this second file would have identical mtime and inode (the inode number just gets recycled right away, it seems). This resulted in false hits from the inode cache, which resulted in subsequent failures later in the build due to configure having given wrong system information.

We had never really intended to use ccache when calling configure, it's just the way things got wired up by setting CC='ccache ...'

So, we're doing two things: getting this small subset of our users migrated to ext4, and avoiding the use of ccache for calls to configure.

There was another case where we also saw this same configure issue, but on ext4. As I understand it, ext4 uses a cached timestamp (from current_fs_time() in the kernel) which is updated at every timer interrupt. So, in our environment, this is:

```
grep 'CONFIG_HZ=' /boot/config-$(uname -r)
CONFIG_HZ=1000
```

so, 1000 times per second, or 1 millisecond resolution. (I was a bit surprised to learn that ext4 doesn't actually get nanosecond resolution -- there's space for it on disk, but at least on Linux, this is limited by the timer interrupt.

But these compiles from the configure shell script certainly weren't happening faster than 1ms (more like 5-20 ms).

What I eventually figured out was that this was caused by some issue with our virtualization setup. I don't yet have a full root cause, but on all our VMs I'm seeing in /var/log/message, warnings like:

```
hpet1  lost 7 rtc interrupts
hpet1  lost 11 rtc interrupts
```

Interestingly, those messages are being logged exactly every 242 seconds, so it's not a simple overloading issue. I was able to demonstrate that every 242 seconds we see a discontinuity in timestamps:

```
# run this to generate a large number of files around the time of the lost interrupts warnings
(n=1; while [[ $n -le 900000 ]]; do > test$n; n=$((n+1)); done)
```

And then search through that to find:

```
ls -ltr --full-time


-rw-r--r-- 1 jamscriv eng   0 2022-12-01 23:39:54.155042412 +0530 test70794
-rw-r--r-- 1 jamscriv eng   0 2022-12-01 23:39:54.155042412 +0530 test70793
-rw-r--r-- 1 jamscriv eng   0 2022-12-01 23:39:54.156042435 +0530 test77435
-rw-r--r-- 1 jamscriv eng   0 2022-12-01 23:39:54.156042435 +0530 test77434
... (6601 "156042435" rows) ...
-rw-r--r-- 1 jamscriv eng   0 2022-12-01 23:39:54.156042435 +0530 test70836
-rw-r--r-- 1 jamscriv eng   0 2022-12-01 23:39:54.156042435 +0530 test70835
-rw-r--r-- 1 jamscriv eng   0 2022-12-01 23:39:54.312046014 +0530 test77474 <--- 156ms discontinuity here
-rw-r--r-- 1 jamscriv eng   0 2022-12-01 23:39:54.312046014 +0530 test77473
```

So, I don't understand what's causing this. I've reached out to Redhat for some help, but in any case for some reason, there are large discontinuities in timestamps even on ext4, on virtual hosts.

Some thoughts I had on how ccache could make this more robust:

1) include the file size in the inode cache key
2) explicitly ignore inode cache for "conftest.c"
3) ignore inode cache if the cached file is newer than 1 or 2 seconds.

regards,

James Scriven
Cisco

**Reply by liljenzin** [reply]
> Some thoughts I had on how ccache could make this more robust:
> 
>     1. include the file size in the inode cache key
> 
>     2. explicitly ignore inode cache for "conftest.c"
> 
>     3. ignore inode cache if the cached file is newer than 1 or 2 seconds.

Thanks for sharing your findings.

(1) is already there. The following struct is filled before each lookup, then the struct is hashed and the hash value is used as key in the hash table lookup:
```
struct InodeCache::Key 
{ 
  ContentType type; 
  dev_t st_dev; 
  ino_t st_ino; 
  mode_t st_mode; 
  timespec st_mtim; 
  timespec st_ctim; // Included for sanity checking. 
  off_t st_size;    // Included for sanity checking. 
}; 
```
Since you still get false hits it must be from files that also have same size.

(2) feels a bit ad hoc. If we can get false hits with this particular filename we can get false hits with other filenames as well and then it would be better to spend the effort on a more generic solution.

(3) sounds like a more generic solution. Bypassing the cache for files newer than a second shouldn't impact overall performance since such files should be rare by obvious reasons and it is unlikely that we would have time to make use of the cached values otherwise.

But (3) alone should be sufficient to fix your issue, right?

Perhaps OT, but a more general thought is if your virtual machines are reliable build hosts by other reasons provided the kernel issue. I was thinking of tools like make that use timestamps for dependency checking. Something like this could also fail, not because ccache is necessarily doing something wrong, but because ccache is not run by make when it should: 
```
conftest.o: conftest.c
  ccache gcc -c conftest.c
```

**Reply by jrosdahl** [reply]
Agreed, thanks for sharing your findings! :+1:

The issue has already been solved in ccache 4.7.4. See #1215 and 3b6c2a5a63e104ed2090ddebe611182297fd1f9c for details.

**Reply by jdscriven** [reply]
I didn't realize size was already in there, so in this case, yes, these files must be identical size, which isn't all that surprising since many very similar checks are performed. Yes, I believe (3) alone would address our issue, and should address even 1-second resolution filesystems.

For sure, I want to get to the bottom of VM time issues, as that is indeed probably causing other subtle issues. For this particular case though, with conftest.c, make isn't actually involved. The one "configure" script is just a shell script which generates and compiles the multiple conftest.c files in sequence.

So putting aside my odd VM behavior, it sounds like a 1 or 2 second delay might be a good idea.

James Scriven
Cisco

**Reply by jdscriven** [reply]
Looks like you're one step ahead of me! Thanks for sharing the 4.7.4 fixes!

James Scriven
Cisco
